client/network/src/service.rs

```diff
@@ -560,6 +568,71 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
 		});
 	}
 
+	/// Appends one or more notifications to the buffer of pending outgoing notifications with the
+	/// given peer.
+	///
+	/// The returned `Future` finishes only after all the notifications have been queued or the
+	/// substream has been closed.
+	///
+	/// This operation is atomic in the sense that either all or none of the notifications have
+	/// been enqueued. In particular, it is guaranteed that no notifications have been transmitted
+	/// if the operation is interrupted by dropping the `Future`.
+	///
+	/// An error is returned if there exists no open notifications substream with that combination
+	/// of peer and protocol, or if the remote has asked to close the notifications substream.
+	///
+	/// If the remote requests to close the notifications substream, all notifications successfully
+	/// enqueued with this method will finish being sent out before the substream actually gets
+	/// closed, but attempting to enqueue more notifications will now return an error. It is also
+	/// possible for the entire connection to be abruptly closed, in which case enqueued
+	/// notifications will be lost.
+	///
+	/// The protocol must have been registered with `register_notifications_protocol` or
+	/// `NetworkConfiguration::notifications_protocols`.
+	///
+	/// # Usage
+	///
+	/// This method waits until there space is available in the buffer of messages towards the
+	/// given peer. If the peer processes notifications at a slower rate than we send them, this
+	/// buffer will quickly fill up.
+	///
+	/// As such, you should never do something like this:
+	///
+	/// ```ignore
+	/// // Do NOT do this
+	/// for peer in peers {
+	///     network.send_notifications(peer, ..., notifications).await;
+	/// }
+	/// ```
+	///
+	/// Doing so would slow down all peers to the rate of the slowest one. A malicious or
+	/// malfunctioning peer could intentionally process notifications at a very slow rate.
+	///
+	/// Instead, you are encouraged to maintain your own buffer of notifications on top of the one
+	/// maintained by `sc-network`, and use `send_notifications` to progressively send out
+	/// elements from your buffer. If this additional buffer is full (which will happen at some
+	/// point if the peer is too slow to process notifications), appropriate measures can be taken,
+	/// such as removing non-critical notifications from the buffer or disconnecting the peer
+	/// using [`NetworkService::disconnect_peer`].
+	///
+	///
+	///  Notifications              Per-peer buffer
+	///    broadcast    +------->   of notifications   +-->  `send_notifications`  +-->  Internet
+	///                     ^       (not covered by
+	///                     |         sc-network)
+	///                     +
+	///       Notifications should be dropped
+	///              if buffer is full
+	///
+	pub async fn send_notifications(
```

Comment:
Is this supposed to mirror my bounded-priority-queue suggestion? If so I am confused by the API description.

The diagram above would seem to suggest that this function is for taking things out of the per-peer buffer, and then sending them to the peer via whatever stream we're using. In other words, it corresponds to `pop_message` as I described in [this comment](https://github.com/paritytech/substrate/issues/5481#issuecomment-654475634).

However a few paragraphs above you say "This method waits until there space is available in the buffer of messages towards", which is not what `pop_message` is supposed to do, it is supposed to wait until *items are available* i.e. the buffer is non-empty - otherwise there is nothing to pop. The corresponding method `push_message` (which in your diagram would be called by the notifications broadcast) is never supposed to block.

Reply:
This function is meant to be the foundation on top of which your bounded-priority-queue could be implemented.
The way they plug together is that you'd do something like `while let Some(next) = pop_message() { send_notifications(..., next) }`

Reply:
Ah ok. It might be good to clarify "there is space available in the **lower-level** buffer of messages towards the peer" and add this to the diagram in between `send_notifications` and `Internet`.